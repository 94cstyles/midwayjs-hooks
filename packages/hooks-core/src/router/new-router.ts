import kebabCase from 'lodash/kebabCase'
import {
  basename,
  dirname,
  extname,
  join,
  relative,
  toUnix,
  removeExt,
} from 'upath'

import { isPathInside, Route } from '..'

export interface RouterConfig {
  root: string
  source: string
  routes: Route[]
}

export class NewFileRouter {
  constructor(public config: RouterConfig) {}

  get source() {
    return join(this.config.root, this.config.source)
  }

  getApiDirectory(baseDir: string) {
    return join(this.source, baseDir)
  }

  getRoute(file: string) {
    return this.config.routes.find((route) => {
      const apiDir = this.getApiDirectory(route.baseDir)
      return isPathInside(toUnix(file), toUnix(apiDir))
    })
  }

  isApiFile(file: string) {
    const extensions = ['.ts', '.tsx', '.js', '.jsx', '.mjs']
    if (!extensions.includes(extname(file))) {
      return false
    }

    const testExt = ['.test.ts', '.test.tsx', '.test.js', '.test.jsx']
    if (testExt.some((ext) => file.endsWith(ext))) {
      return false
    }

    const route = this.getRoute(file)
    return !!route
  }

  getFunctionId(file: string, functionName: string, isExportDefault: boolean) {
    const relativePath = relative(this.source, file)
    // src/apis/lambda/index.ts -> apis-lambda-index
    const id = kebabCase(removeExt(relativePath, extname(relativePath)))
    const name = [id, isExportDefault ? '' : `-${functionName}`].join('')
    return name.toLowerCase()
  }

  fileToHttpPath(file: string, functionName: string, exportDefault: boolean) {
    const { basePath, baseDir } = this.getRoute(file)

    const { isCatchAllRoutes, filename } = this.parseFilename(
      basename(file, extname(file))
    )

    const dirPath = relative(this.getApiDirectory(baseDir), dirname(file))

    const api = toUnix(
      join(
        basePath,
        /**
         * /apis/lambda/index.ts -> ''
         * /apis/lambda/note/index.ts -> 'note'
         */
        dirPath,
        /**
         * index -> ''
         * demo -> '/demo'
         */
        filename === 'index' ? '' : filename,
        exportDefault ? '' : functionName,
        isCatchAllRoutes ? '/*' : ''
      )
    )

    return api
  }

  private parseFilename(filename: string) {
    const re = /\[\.{3}(.+)]/
    const isCatchAllRoutes = re.test(filename)

    return {
      isCatchAllRoutes,
      filename: isCatchAllRoutes ? re.exec(filename)?.[1] : filename,
    }
  }
}
