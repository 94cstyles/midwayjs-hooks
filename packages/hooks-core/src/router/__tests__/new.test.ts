import { NewFileRouter } from '../new-router'

describe('NewFileRouter', () => {
  let router: NewFileRouter

  beforeEach(() => {
    router = new NewFileRouter({
      root: '/',
      source: 'src',
      routes: [
        {
          baseDir: 'lambda',
          basePath: '/api',
        },
      ],
    })
  })

  test('should exist', () => {
    expect(NewFileRouter).toBeTruthy()
    expect(router).toBeInstanceOf(NewFileRouter)
  })

  test('test file router', () => {
    expect(router.source.endsWith('/src')).toBeTruthy()

    const api = '/src/lambda/index.ts'
    expect(router.isApiFile(api)).toBeTruthy()
    expect(router.isApiFile('/dist')).toBeFalsy()

    const rule = router.getRoute(api)
    expect(rule).toMatchSnapshot()
    expect(
      router.getApiDirectory(rule.baseDir).endsWith('src/lambda')
    ).toBeTruthy()
  })
})
