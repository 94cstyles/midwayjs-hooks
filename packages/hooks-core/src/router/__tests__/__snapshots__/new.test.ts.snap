// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewFileRouter test file router 1`] = `
Object {
  "baseDir": "lambda",
  "basePath": "/api",
}
`;
